filters/auth/grantcookie.go

```diff
@@ -0,0 +1,126 @@
+package auth
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"net/http"
+	"time"
+
+	"github.com/zalando/skipper/secrets"
+	"golang.org/x/oauth2"
+)
+
+type cookie struct {
+	AccessToken  string    `json:"access_token"`
+	RefreshToken string    `json:"refresh_token"`
+	Expiry       time.Time `json:"expiry,omitempty"`
+	RefreshAfter time.Time `json:"refresh_after,omitempty"`
+}
+
+func refreshAfter(expiry time.Time) time.Time {
+	now := time.Now()
+	d := expiry.Sub(now)
+	if d <= 0 {
+		return now
+	}
+
+	d /= 10
+	if d < time.Minute {
+		d = time.Minute
+	}
+
+	return now.Add(d)
+}
+
+func decodeCookie(cookieHeader string, config OAuthConfig) (c *cookie, err error) {
+	var eb []byte
+	if eb, err = base64.StdEncoding.DecodeString(cookieHeader); err != nil {
+		return
+	}
+
+	var encryption secrets.Encryption
+	if encryption, err = config.Secrets.GetEncrypter(secretsRefreshInternal, config.SecretFile); err != nil {
+		return
+	}
+
+	var b []byte
+	if b, err = encryption.Decrypt(eb); err != nil {
+		return
+	}
+
+	err = json.Unmarshal(b, &c)
+	return
+}
+
+func (c cookie) isAccessTokenExpired() bool {
+	now := time.Now()
+	return now.After(c.Expiry)
+}
+
+// GetCookie extracts the OAuth Grant token cookie from a HTTP request.
+// The function supports multiple cookies with the same name and returns
+// the best match (the one that decodes properly).
+// The client may send multiple cookies if a parent domain has set a
+// cookie of the same name.
```

Comment:
Note: seems to be fine, because cookie size is limited. If not it would be cheaper to attack the cookie for attackers